src/core/ddsi/src/ddsi_plist.c

```diff
@@ -678,6 +681,104 @@ static bool print_data_representation (char * __restrict *buf, size_t * __restri
   return prtf (buf, bufsize, ")");
 }
 
+#ifdef DDS_HAS_TYPE_DISCOVERY
+
+static dds_return_t deser_type_information (void * __restrict dst, struct flagset *flagset, uint64_t flag, const struct dd * __restrict dd, struct ddsi_domaingv const * const gv)
+{
+  (void) gv;
+  size_t dstoff = 0;
+  uint32_t srcoff = 0;
+  unsigned char *buf;
+  dds_return_t ret = 0;
+
+  if (dd->bswap)
+  {
+    buf = ddsrt_memdup (dd->buf, dd->bufsz);
+    if (!dds_stream_normalize_data ((char *) buf, &srcoff, (uint32_t) dd->bufsz, dd->bswap, CDR_ENC_VERSION_2, DDS_XTypes_TypeInformation_desc.m_ops))
+    {
+      ret = DDS_RETCODE_BAD_PARAMETER;
+      goto err_normalize;
+    }
+  }
+  else
+    buf = (unsigned char *) dd->buf;
+
+  dds_istream_t is = { .m_buffer = buf, .m_index = 0, .m_size = (uint32_t) dd->bufsz, .m_xcdr_version = CDR_ENC_VERSION_2 };
+  ddsi_typeinfo_t const ** x = deser_generic_dst (dst, &dstoff, alignof (ddsi_typeinfo_t *));
+  *x = ddsrt_calloc (1, DDS_XTypes_TypeInformation_desc.m_size);
+  dds_stream_read (&is, (void *) *x, DDS_XTypes_TypeInformation_desc.m_ops);
+  *flagset->present |= flag;
+err_normalize:
+  if (dd->bswap)
+    ddsrt_free (buf);
+  return ret;
+}
+
+static dds_return_t ser_type_information (struct nn_xmsg *xmsg, nn_parameterid_t pid, const void *src, size_t srcoff, enum ddsrt_byte_order_selector bo)
+{
+  ddsi_typeinfo_t const * const * x = deser_generic_src (src, &srcoff, alignof (ddsi_typeinfo_t *));
+
+  dds_ostream_t os = { .m_buffer = NULL, .m_index = 0, .m_size = 0, .m_xcdr_version = CDR_ENC_VERSION_2 };
+  if (bo == DDSRT_BOSEL_LE)
```

Comment:
I'd find it more elegant if there'd be a `dds_stream_write_with_byte_order` or so. It only moves the cases here to another place, but it feels better to me.

Reply:
Fixed in adb9c5e